src/main/java/htsjdk/samtools/util/EdgeReadIterator.java

```diff
@@ -0,0 +1,227 @@
+    void accumulateSamRecord(SAMRecord rec) {
+        // interpret the CIGAR string and add the base info
+        for (final AlignmentBlock alignmentBlock : rec.getAlignmentBlocks()) {
+            // 0-based offset into the read of the current base
+            int readOffset = alignmentBlock.getReadStart() - 1;
+            // 1-based reference position that the current base aligns to
+            int refPos = alignmentBlock.getReferenceStart();
+
+            // 0-based offset from the aligned position of the first base in the read to the aligned position
+            // of the current base.
+            int refOffset = refPos - rec.getAlignmentStart();
+            int refOffsetEnd = refPos - rec.getAlignmentStart() + alignmentBlock.getLength();
+
+
+            // Ensure there are AbstractLocusInfos up to and including this position
+            for (int j = accumulator.size(); j <= refOffsetEnd; ++j) {
+                accumulator.add(createLocusInfo(getReferenceSequence(rec.getReferenceIndex()),
+                        rec.getAlignmentStart() + j));
+            }
+
+            /* Let's assume an alignment block starts in some locus. 
+             * We put two records to the accumulator. The first one has the "begin" type which corresponds to the locus 
+             * where the block starts. The second one has the "end" type which corresponds to the other locus where the block ends. 
+            */
+            int refOffsetInterval = refOffset; // corresponds to the beginning of the alignment block 
+            int refOffsetEndInterval = refOffsetEnd;
+            int startShift = 0;
+
+            // intersect intervals and alignment block
+            if (getIntervals() != null) {
+                // get the current interval we're proceeding 
+                Interval interval = getCurrentInterval();
+                if (interval != null) {
+                    final int intervalEnd = interval.getEnd();
+                    final int intervalStart = interval.getStart();
+                    // check if an interval and the alignment block overlap 
+                    if ((refPos < intervalStart && refPos + alignmentBlock.getLength() < intervalStart)
+                            || (refPos > intervalEnd && refPos + alignmentBlock.getLength() > intervalEnd)) {
+                        continue;
+                    }
+                    // if the alignment block starts out of an interval shift the starting position
+                    if (refPos < intervalStart) {
+                        startShift = intervalStart - refPos;
+                        refOffsetInterval = refOffsetInterval + startShift;
+                    }
+                    // if the alignment block ends out of an interval shift the ending position
+                    if (refPos + alignmentBlock.getLength() > intervalEnd) {
+                        refOffsetEndInterval = refOffsetEndInterval - (refPos + alignmentBlock.getLength() - intervalEnd) + 1;
```

Comment:
These lines are not equivalent, I've tried several variants, but unfortunately I failed to make it final. I hope at least the comment clarifies the code.